File: src/org/newtonpolyhedron/solve/polyinter/PolyIntersectionSolverImpl.java
package org.newtonpolyhedron.solve.polyinter;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.NoSuchElementException;

import org.fs.utils.ListUtils;
import org.fs.utils.collection.iter.AbstractIterator;
import org.fs.utils.collection.map.BasicSortedMap;
import org.newtonpolyhedron.entity.vector.FractionVector;
import org.newtonpolyhedron.entity.vector.IntVector;
import org.newtonpolyhedron.solve.cone.ConeSolver;
import org.newtonpolyhedron.utils.ArithUtils;
import org.newtonpolyhedron.utils.NullPrintWriter;
import org.newtonpolyhedron.utils.PointUtils;

public class PolyIntersectionSolverImpl implements PolyIntersectionSolver {
	
	private final ConeSolver	coneSolver;
	
	public PolyIntersectionSolverImpl(final ConeSolver coneSolver) {
		this.coneSolver = coneSolver;
	}
	
	@Override
	public Map <IntVector, List <List <Integer>>> solve(
			final List <List <FractionVector>> polyhedrons,
			final int dim) throws Exception {
		if (dim < 3)
			throw new IllegalArgumentException("No intersections are possible below 3-dimension");
		if (polyhedrons.size() < dim - 1)
			throw new IllegalArgumentException(
					"Not enough polyhedrons for intersecting at dimension " + dim);
		final List <Integer> polyPtsCount = new ArrayList <Integer>();
		for (final List <FractionVector> poly : polyhedrons) {
			polyPtsCount.add(poly.size());
		}
		
		final Map <IntVector, List <List <Integer>>> ptsForVectors = new BasicSortedMap <IntVector, List <List <Integer>>>();
		final int polyCount = polyhedrons.size();
		
		for (final List <Integer> indices : new IndicesIterable(polyCount, polyPtsCount)) {
			
			// ++ Construct a system
			final List <List <IntVector>> eqSystems = new ArrayList <List <IntVector>>();
			for (int i = 0; i < polyCount; ++i) {
				eqSystems.add(PointUtils.copySubtractPointAsInt(polyhedrons.get(i), indices.get(i)));
			}
			final List <IntVector> commonEqSys = new ArrayList <IntVector>();
			for (final List <IntVector> eqSys : eqSystems) {
				commonEqSys.addAll(eqSys);
			}
			// -- Construct a system
			
			final List <IntVector> solution = coneSolver.solve(commonEqSys, null, dim,
					NullPrintWriter.instance);
			
			removeNonIntersectingSolutions(solution, eqSystems);
			
			for (final IntVector solutionVector : solution) {
				List <List <Integer>> list = ptsForVectors.get(solutionVector);
				if (list == null) {
					list = new ArrayList <List <Integer>>();
					ptsForVectors.put(solutionVector, list);
				}
				list.add(indices);
			}
		}
		return ptsForVectors;
	}
	
	protected static void removeNonIntersectingSolutions(
			final List <IntVector> solution,
			final List <List <IntVector>> equationSystems) {
		for (final Iterator <IntVector> solutionIter = solution.iterator(); solutionIter.hasNext();) {
			final IntVector currSolution = solutionIter.next();
			for (final List <IntVector> currEqSystem : equationSystems) {
				boolean nullifiesAtLeastOne = false;
				for (final IntVector eq : currEqSystem) {
					if (ArithUtils.isZero(eq.dotProduct(currSolution))) {
						nullifiesAtLeastOne = true;
						break;
					}
				}
				if (!nullifiesAtLeastOne) {
					solutionIter.remove();
					break;
				}
			}
		}
	}
	
	//
	// Inner classes
	//
	private static class IndicesIterable implements Iterable <List <Integer>> {
		
		private final int				length;
		private final List <Integer>	maximums;
		
		public IndicesIterable(final int length, final List <Integer> maximums) {
			this.length = length;
			this.maximums = maximums;
			
		}
		
		@Override
		public Iterator <List <Integer>> iterator() {
			return new IndicesIterator(length);
		}
		
		private class IndicesIterator extends AbstractIterator <List <Integer>> {
			
			private List <Integer>	current;
			private List <Integer>	next;
			
			public IndicesIterator(final int length) {
				current = new ArrayList <Integer>(//
						Collections.nCopies(//
								length,//
								Integer.valueOf(0)//
						)//
				);
				current.set(length - 1, -1);
			}
			
			@Override
			public synchronized boolean hasNext() {
				if (next != null) return true;
				doGetNext();
				return next != null;
			}
			
			@Override
			public synchronized List <Integer> next() {
				if (!hasNext()) throw new NoSuchElementException();
				current = next;
				next = null;
				return new ArrayList <Integer>(current);
			}
			
			private void doGetNext() {
				next = new ArrayList <Integer>(current);
				int idx = next.size() - 1;
				ListUtils.inc(next, idx);
				while (idx > 0) {
					if (next.get(idx) < maximums.get(idx)) {
						break;
					}
					next.set(idx, 0);
					--idx;
					ListUtils.inc(next, idx);
				}
				if (next.get(0) >= maximums.get(0)) {
					next = null;
				}
			}
		}
	}
}

File: src/main/scala/org/newtonpolyhedron/solve/polyinter/PolyIntersectionSolverImpl.scala
package org.newtonpolyhedron.solve.polyinter

import scala.collection.immutable.SortedSet

import org.fs.utility.collection.table.KeyTable
import org.newtonpolyhedron.NewtonImports._
import org.newtonpolyhedron.solve.cone.ConeSolver
import org.newtonpolyhedron.utils.PointUtils

class PolyIntersectionSolverImpl[N <: MPNumber](val coneSolver: ConeSolver)(implicit mp: MathProcessor[N])
    extends PolyIntersectionSolver[N] {

  override def solve(
      polyhedrons: Seq[IndexedSeq[NumVec[N]]],
      dim:         Int
  ): KeyTable[Int, IntVec, SortedSet[Int]] = {
    require(dim >= 3, "No intersections are possible below 3-dimension")
    require(polyhedrons.size >= dim - 1, "Not enough polyhedrons for intersecting at dimension " + dim)
    val polySizes = polyhedrons.map(_.size).toIndexedSeq
    val initialIndices = IndexedSeq.fill(polyhedrons.size)(0).updated(polyhedrons.size - 1, -1)
    val indices = indicesStream(initialIndices, polySizes)
    val ptsForVectors = fillPtsForVectorsMap(indices, polyhedrons, dim)
    val ptsForPolysAndVectors = reverseTableMeaning(ptsForVectors)
    ptsForPolysAndVectors
  }

  /** @return { vector : [ point indices list per polyhedron ] } */
  def fillPtsForVectorsMap(
      indicesSeq:  Seq[IndexedSeq[Int]],
      polyhedrons: Seq[IndexedSeq[NumVec[N]]],
      dim:         Int
  ): Map[IntVec, Seq[IndexedSeq[Int]]] = {
    def fillMapRecursive(
        indicesSeq: Seq[IndexedSeq[Int]],
        prevMap:    Map[IntVec, Seq[IndexedSeq[Int]]]
    ): Map[IntVec, Seq[IndexedSeq[Int]]] =
      if (indicesSeq.isEmpty) prevMap
      else {
        val indices = indicesSeq.head
        // ++ Construct a system
        val eqSystems = polyhedrons mapWithIndex { (poly, i) =>
          val idx = indices(i)
          PointUtils.copySubtractPointAsInt(poly, idx)
        }
        val commonEqSys = eqSystems.flatten
        // -- Construct a system

        val solutions = coneSolver solve (commonEqSys, None, dim)
        val intersectingSols = solutions filter (isIntersectingSol(eqSystems))
        fillMapRecursive(indicesSeq.tail, withSoluionsToIndices(prevMap, intersectingSols, indices))
      }
    fillMapRecursive(indicesSeq, Map.empty)
  }

  def withSoluionsToIndices(
      prevMap:   Map[IntVec, Seq[IndexedSeq[Int]]],
      solutions: Seq[IntVec],
      indices:   IndexedSeq[Int]
  ): Map[IntVec, Seq[IndexedSeq[Int]]] =
    solutions.headOption match {
      case None => prevMap
      case Some(sol) => {
        val nextMap = if (prevMap contains sol)
          prevMap + (sol -> (prevMap(sol) :+ indices))
        else
          prevMap + (sol -> IndexedSeq(indices))
        withSoluionsToIndices(nextMap, solutions.tail, indices)
      }
    }

  def indicesStream(
      curr:     IndexedSeq[Int],
      maximums: IndexedSeq[Int]
  ): Stream[IndexedSeq[Int]] = {
    def calcNextRec(
        idx:      Int,
        curr:     IndexedSeq[Int],
        maximums: IndexedSeq[Int]
    ): IndexedSeq[Int] = {
      val prev = idx - 1
      if (idx == 0 || curr(idx) < maximums(idx)) curr
      else calcNextRec(idx - 1, curr.updated(idx, 0).updated(prev, curr(prev) + 1), maximums)
    }
    val len = curr.length
    val preNextSeq = curr updated (len - 1, curr.last + 1)
    val nextSeq = calcNextRec(len - 1, preNextSeq, maximums)
    if (nextSeq(0) < maximums(0))
      nextSeq #:: indicesStream(nextSeq, maximums)
    else Stream.empty
  }

  def isIntersectingSol(equationSystems: Seq[Seq[IntVec]])(solution: IntVec): Boolean =
    equationSystems forall (_ exists (_ *+ solution == 0))

  /**
   * @param ptsForVectors vector -> [ point indices list per polyhedron ]
   * @return (polyIdx, vector) -> [ points giving this vector for this poly when intersecting ]
   */
  def reverseTableMeaning(ptsForVectors: Map[IntVec, Seq[Seq[Int]]]): KeyTable[Int, IntVec, SortedSet[Int]] = {
    val vectPtTable =
      ptsForVectors.foldLeft(KeyTable.empty[Int, IntVec, SortedSet[Int]]) {
        case (table, (vector, ptIndicesSeq)) => {
          ptIndicesSeq.foldLeft(table) { (table, ptIndices) =>
            ptIndices.zipWithIndex.foldLeft(table) {
              case (table, (ptIndex, i)) =>
                val pts = table.get(i, vector) getOrElse SortedSet.empty[Int]
                table + (i, vector, pts + ptIndex)
            }
          }
        } ensuring (table => {
          // Sanity check that all combinations are present
          // I.e. [0, 5], [0, 6], [0, 7], [1, 5], [1, 6], [1, 7] must all present to get [0, 1], [5, 6, 7]
          // We'll use a weaker assertion: number of combinations must be equal to product of resulting point set sizes
          table.col(vector).values.map(_.size).product == ptIndicesSeq.size
        }, "Logic failure, please report to developer")
      }
    vectPtTable.sortedCols
  }
}
